### PR TITLE
CustomCommandManager + CustomCommandManagerSession to use auto-expandable maps

### PR DESCRIPTION
**Pull request by TalZaccai**
- [x] Implementing `ExpandableMap` struct which maps a specified range of IDs to an array which doubles in size as needed.
- [x] Switching `CustomCommandManager` & `CustomCommandManagerSession` to use `ExpandableMap` instead of using set-sized arrays.
- [x] Adding tests to `RespCustomCommandTests` to test registration of multiple string commands, object commands, transactions and procedures.

Note: this relates to closed issue #834, which was addressed by temporary fix in PR #836.

Benchmark results:

main:
| Method                 | Params | Mean       | Error    | StdDev   | Gen0   | Allocated |
|----------------------- |------- |-----------:|---------:|---------:|-------:|----------:|
| CustomRawStringCommand | ACL    |   409.1 us |  7.60 us |  7.11 us |      - |         - |
| CustomObjectCommand    | ACL    | 1,061.6 us | 17.22 us | 16.11 us | 1.9531 |   24002 B |
| CustomTransaction      | ACL    | 1,738.7 us | 31.31 us | 49.66 us |      - |       2 B |
| CustomProcedure        | ACL    | 1,264.8 us | 14.86 us | 13.90 us |      - |       2 B |
| CustomRawStringCommand | AOF    |   414.0 us |  4.80 us |  4.49 us |      - |         - |
| CustomObjectCommand    | AOF    | 1,227.9 us | 16.35 us | 15.30 us | 1.9531 |   24002 B |
| CustomTransaction      | AOF    | 1,967.2 us |  6.12 us |  5.43 us |      - |       4 B |
| CustomProcedure        | AOF    | 1,678.9 us | 27.81 us | 38.99 us |      - |       2 B |
| CustomRawStringCommand | None   |   403.5 us |  7.81 us |  7.31 us |      - |         - |
| CustomObjectCommand    | None   | 1,110.4 us | 21.81 us | 31.28 us | 1.9531 |   24002 B |
| CustomTransaction      | None   | 1,732.3 us | 12.80 us |  9.99 us |      - |       2 B |
| CustomProcedure        | None   | 1,229.2 us | 24.41 us | 21.64 us |      - |       2 B |

PR #848:
| Method                 | Params | Mean       | Error    | StdDev   | Gen0   | Allocated |
|----------------------- |------- |-----------:|---------:|---------:|-------:|----------:|
| CustomRawStringCommand | ACL    |   411.1 us |  1.74 us |  1.63 us |      - |         - |
| CustomObjectCommand    | ACL    | 1,105.2 us |  2.20 us |  1.95 us | 1.9531 |   24002 B |
| CustomTransaction      | ACL    | 1,668.7 us | 20.38 us | 19.07 us |      - |       2 B |
| CustomProcedure        | ACL    | 1,205.2 us |  1.04 us |  0.93 us |      - |       2 B |
| CustomRawStringCommand | AOF    |   394.8 us |  1.59 us |  1.33 us |      - |         - |
| CustomObjectCommand    | AOF    | 1,221.4 us |  3.30 us |  2.92 us | 1.9531 |   24002 B |
| CustomTransaction      | AOF    | 1,861.3 us |  3.48 us |  3.08 us |      - |       2 B |
| CustomProcedure        | AOF    | 1,632.2 us | 24.00 us | 22.45 us |      - |       2 B |
| CustomRawStringCommand | None   |   402.2 us |  1.39 us |  1.30 us |      - |         - |
| CustomObjectCommand    | None   | 1,108.0 us |  1.65 us |  1.46 us | 1.9531 |   24002 B |
| CustomTransaction      | None   | 1,646.6 us |  5.77 us |  5.11 us |      - |       2 B |
| CustomProcedure        | None   | 1,209.7 us |  2.49 us |  1.95 us |      - |       2 B |

### CODE REVIEW

File: libs/server/Custom/ExpandableMap.cs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    {
+        /// <summary>
+        /// The underlying array containing the items
+        /// </summary>
+        internal T[] map;
+
+        /// <summary>
+        /// Reader-writer lock for the underlying item array
+        /// </summary>
+        internal readonly ReaderWriterLockSlim mapLock = new();
```
Comment (badrishc) [inline libs/server/Custom/ExpandableMap.cs:24]: Can we use @vazois's `SingleWriterMultiReaderLock`?

File: libs/server/Custom/ExpandableMap.cs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+        /// <summary>
+        /// Creates a new instance of ExpandableMap
+        /// </summary>
+        /// <param name="minSize">Initial size of underlying array</param>
+        /// <param name="minId">The minimal item ID value</param>
+        /// <param name="maxId">The maximal item ID value (can be smaller than minId for descending order of IDs)</param>
+        public ExpandableMap(int minSize, int minId, int maxId)
+        {
+            this.map = [];
```
Comment (badrishc) [inline libs/server/Custom/ExpandableMap.cs:50]: can this start as null so we have no allocations until necessary?

File: libs/server/Custom/CustomCommandManagerSession.cs [hunk 3]
```diff
@@ -13,53 +14,81 @@ internal sealed class CustomCommandManagerSession
...
-
+        ExpandableMap<CustomTransactionProcedureWithArity> sessionTransactionProcMap;
+        ExpandableMap<CustomProcedure> sessionCustomProcMap;
 
         public CustomCommandManagerSession(CustomCommandManager customCommandManager)
         {
             this.customCommandManager = customCommandManager;
-            sessionTransactionProcMap = new (CustomTransactionProcedure, int)[CustomCommandManager.MaxRegistrations];
-            sessionCustomProcMap = new CustomProcedure[CustomCommandManager.MaxRegistrations];
+            sessionTransactionProcMap = new ExpandableMap<CustomTransactionProcedureWithArity>(CustomCommandManager.MinMapSize, 0, byte.MaxValue);
```
Comment (badrishc) [inline libs/server/Custom/CustomCommandManagerSession.cs:23]: Why does CustomCommandManagerSession need to pay the cost of locking the ExpandableMap - the idea of creating a per session copy was to make the per-session critical path blazingly fast and lock free.

Comment (TalZaccai) [reply libs/server/Custom/CustomCommandManagerSession.cs:23]: Ah I see what you're saying, makes sense.

File: libs/server/Custom/ExpandableMap.cs (hunk 1 above)
Comment (TalZaccai) [reply libs/server/Custom/ExpandableMap.cs:24]: I don't think it supports upgrading an acquired read lock to a write lock, right @vazois?

Comment (badrishc) [reply libs/server/Custom/ExpandableMap.cs]: why does it need upgradability? this might be more efficient than what C# provides.

Comment (TalZaccai) [reply libs/server/Custom/ExpandableMap.cs]: Thought about it, it's fine as is :)

File: libs/server/Custom/CustomCommandManagerSession.cs (hunk 3 above)
Comment (badrishc) [inline libs/server/Custom/CustomCommandManagerSession.cs:23]: how does this get updated when procedures are added to the main structure? see how this is handled in `main` - we might need a similar approach.

Comment (TalZaccai) [reply libs/server/Custom/CustomCommandManagerSession.cs:23]: The session map is a "cache" of the main map, if the id is not in the session map then it checks the main map and updates the session map.

**Review by badrishc — APPROVED** [review]

**Pull request merged**